deltachat-ios/AppDelegate.swift

```diff
@@ -372,38 +372,45 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
         bgIoTimestamp = nowTimestamp
 
         // make sure to balance each call to `beginBackgroundTask` with `endBackgroundTask`
+        let fetchSemaphore = DispatchSemaphore(value: 0)
         let backgroundTask = UIApplication.shared.beginBackgroundTask {
-            // TODO: currently, this should be okay, we are not using too much background time,
-            // so that handler should never be called.
-            // the plan is to listen to an event as DC_EVENT_ENTER_IDLE and stop fetch from there.
-            // if we have such an event, we could imprive this handler and fire the event.
-            logger.info("fetch background task will end soon")
+            // we do not call `endBackgroundTask` here, even if that results in a warning:
+            // if we would do, we would risk a 0xdead10cc exception as things are still running;
+            // therefore we just signal the task below to exit gracefully and call `endBackgroundTask` from there.
+            // in practise, this handler should not be called anyway as we take care to stay below the 30 seconds.
+            logger.info("⬅️ background task handler called, finishing fetch")
+            fetchSemaphore.signal()
         }
 
         // we're in background, run IO for a little time
         dcContext.maybeStartIo()
         dcContext.maybeNetwork()
 
-        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
-            logger.info("⬅️ finishing fetch")
+        DispatchQueue.global(qos: .background).async { [weak self] in
+            logger.info("➡️ fetching")
+            _ = fetchSemaphore.wait(timeout: .now() + 1000)
```

Comment:
The initializer of DispatchTime, the argument passed to the wait function, uses nano-seconds https://developer.apple.com/documentation/dispatch/dispatchtime. 
Did you test that + 1000 is right here? You're intending 10 seconds, right?